File: print-server/src/print_server/printer.py
import textwrap

class Printer:
    def __init__(self, device='/dev/usb/lp0', encoding='cp437'):
        self.device = device
        self.encoding = encoding
        self.line_width = 64

    def _write(self, data: bytes):
        with open(self.device, 'wb') as printer:
            printer.write(data)

    def _write_line(self, line: str):
        self._write(line.encode(self.encoding, errors='replace') + b'\n')

    def format_line(self, line: str, align: str = "left") -> str:
        if align == "center":
            return line.center(self.line_width)
        elif align == "right":
            return line.rjust(self.line_width)
        else:
            return line.ljust(self.line_width)

    def text(self, message: str, align: str = "left"):
        for raw_line in message.split('\n'):
            wrapped_lines = textwrap.wrap(raw_line, self.line_width)
            for line in wrapped_lines or ['']:
                self._write_line(self.format_line(line, align))

    def blank_line(self, count=1):
        self._write(b'\n' * count)

    def cut(self):
        self.blank_line(6)
        self._write(b'\x1D\x56\x00')
        
    def _print_qrcode(self, data: str):
        qr_data = data.encode('utf-8')
        length = len(qr_data) + 3
        len_low = length & 0xFF
        len_high = (length >> 8) & 0xFF

        cmds = bytearray()
        cmds += b'\x1d\x28\x6b\x04\x00\x31\x41\x32\x00'
        cmds += b'\x1d\x28\x6b\x03\x00\x31\x43\x06'
        cmds += b'\x1d\x28\x6b\x03\x00\x31\x45\x30'
        cmds += b'\x1d\x28\x6b' + bytes([len_low, len_high]) + b'\x31\x50\x30' + qr_data
        cmds += b'\x1d\x28\x6b\x03\x00\x31\x51\x30'
        cmds += b'\n'
        
        self._write(cmds)

    def print_job(self, job):
        for item in job:
            type_ = item.get("type")
            if type_ == "text":
                self.text(item.get("content", ""), align=item.get("align", "left"))
            elif type_ == "blank":
                self.blank_line(item.get("count", 1))
            elif type_ == "line":
                self.text("-" * self.line_width)
            elif type_ == "cut":
                self.cut()
            elif type_ == "qrcode":
                self._print_qrcode(item.get("content", ""))

File: bearprint-api/localprinter/printer.go
package localprinter

import (
	"io"
	"strings"
	"unicode/utf8"

	"github.com/ian-antking/bear-print/shared/printer"
)

type Printer struct {
	encoding  string
	lineWidth int
	writer    io.Writer
}

func NewPrinter(writer io.Writer) *Printer {
	return &Printer{
		writer:    writer,
		encoding:  "cp437",
		lineWidth: 48,
	}
}

func (p *Printer) write(data []byte) error {
	_, err := p.writer.Write(data)
	return err
}

func (p *Printer) writeLine(line string) error {
	return p.write(append([]byte(line), '\n'))
}

func (p *Printer) formatLine(line string, align printer.Alignment) string {
	pad := p.lineWidth - utf8.RuneCountInString(line)
	if pad <= 0 {
		return line
	}
	switch align {
	case "center":
		left := pad / 2
		right := pad - left
		return strings.Repeat(" ", left) + line + strings.Repeat(" ", right)
	case "right":
		return strings.Repeat(" ", pad) + line
	default:
		return line + strings.Repeat(" ", pad)
	}
}

func (p *Printer) Text(item printer.PrintItem) error {
  lines := strings.Split(item.Content, "\n")
  for _, rawLine := range lines {
    wrapped := wrapText(rawLine, p.lineWidth)
    if len(wrapped) == 0 {
      wrapped = []string{""}
    }
    for _, line := range wrapped {
      err := p.writeLine(p.formatLine(line, item.Align))
      if err != nil {
        return err
      }
    }
  }
  return nil
}

func (p *Printer) BlankLine(count int) error {
	return p.write([]byte(strings.Repeat("\n", count)))
}

func (p *Printer) Cut() error {
	err := p.BlankLine(6)
	if err != nil {
		return err
	}
	return p.write([]byte{0x1D, 0x56, 0x00})
}

func (p *Printer) printQRCode(data string, align printer.Alignment) error {
	cmds := buildQRCodeCmd(data, align, p.lineWidth)
	return p.write(cmds)
}

func (p *Printer) PrintJob(items []printer.PrintItem) error {
  for _, item := range items {
    switch item.Type {
    case "text":
      if err := p.Text(item); err != nil {
        return err
      }
    case "blank":
      if err := p.BlankLine(item.Count); err != nil {
        return err
      }
    case "line":
      lineItem := printer.PrintItem{Content: strings.Repeat("-", p.lineWidth), Align: "left"}
      if err := p.Text(lineItem); err != nil {
        return err
      }
    case "cut":
      if err := p.Cut(); err != nil {
        return err
      }
    case "qrcode":
      if err := p.printQRCode(item.Content, item.Align); err != nil {
        return err
      }
    }
  }
  return nil
}
